### DISCUSSION TITLE
`>` and `|` doesn't work with alias.

### DISCUSSION THREAD

**Discussion by LinArcX** [leo-arch/clifm#280] [Q&A] [answered]
when i try these commands inside clifm, they work perfectly:
```
ls | wc -l
ls > test.txt
```

I expect the same thing for alias internal command of clifm, but it doesn't work. why?
```
N1T1[linarcx@gnu_linux] ~/VoidConf $ alias | wc -l
clifm: '|': No such alias
N1T1[linarcx@gnu_linux] ~/VoidConf $ alias > wc -l
clifm: '>': No such alias
```

I didn't test other sub-commands and I don't know if this is a general behaviour or not.

Is there any workaround for it?

**Comment by LinArcX** [comment]
Interesting. For `actions | wc -l`, it doesn't throw an error. it shows the help page of action:
```
T1[linarcx@gnu_linux] ~ $ actions | wc -l
List or edit actions/plugins

USAGE
  actions [list | edit [APP]]

EXAMPLES
- List available actions/plugins
    actions list
  Note: Since 'list' is the default action, it can be omitted
- Open/edit the actions configuration file with nano
    actions edit nano
- Open/edit the actions configuration file with the default application
    actions edit

Actions are just names for plugins.
An action definition has the following form: "NAME=plugin", for example:
"//=rgfind.sh".
To run a plugin enter the action name. For example, to run the rgfind.sh plugin,
enter "//".
Some plugins accept parameters. To get information about a specific plugin
use the -h,--help flag. Example: "- --help".
```

**Comment by leo-arch** [comment, marked as answer]
Hi @LinArcX.

>  expect the same thing for alias internal command of clifm, but it doesn't work. why?

This precisely because `alias` is an **internal** command. **Clifm** by itself does not support pipes or stream redirection, so that it is taking `|` as a parameter for the `alias` command, and since there's no such alias, it throws the appropriate error.

> For actions | wc -l, it doesn't throw an error. it shows the help page of action

In this case there's not error message, since `actions` does not take any parameter (they're just ignored).

>  didn't test other sub-commands and I don't know if this is a general behaviour or not.

This is the case for **all** internal commands. Since we do not support pipes or stream redirection, every time these shell functions are detected, and no internal command is involved, **clifm** sends the whole command to the system shell (which is why `ls | wc -l` works as expected -there's not internal command involved). Why not to do this for internal commands as well? Simply because the shell does not know anything about **clifm**'s internal commands.

> Is there any workaround for it?

Yes, but I'm pretty sure this is not what you're looking for. Since actions and aliases (just like most config stuff) are stored in text files (which as such are accessible from outside **clifm**), it is always possible to operate on these files using shell commands. Example:

```sh
# Count number of aliases
grep ^alias $CLIFM/clifmrc | wc -l
```
```sh
# Count number of actions
grep -Ev "^#|^$" $CLIFM/actions.clifm | wc -l
```

This is a common procedure for writing plugins, but again, this is most likely not what you want. 

Note: The **CLIFM** environment variable holds the absolute path to your current **clifm** profile directory.

**Reply by LinArcX** [reply]
awesome explanation. thanks a lot.
